src/jailer/src/cgroup.rs

```diff
@@ -1,25 +1,224 @@
+pub struct CgroupBuilder {
+    version: CgroupVersion,
+    hierarchies: HashMap<String, PathBuf>,
+    mount_points: Vec<CgroupMountPoint>,
+}
+
+impl CgroupBuilder {
+    // Creates the builder object
+    // It will discover cgroup mount points and hierarchies configured
+    // on the system and cache the info required to create cgroups later
+    // within this hierarchies
+    pub fn new(ver: CgroupVersion) -> Result<Self> {
+        let mut b = CgroupBuilder {
+            version: ver,
+            hierarchies: HashMap::new(),
+            mount_points: Vec::new(),
+        };
+
+        // search PROC_MOUNTS for cgroup mount points
+        let f =
+            File::open(PROC_MOUNTS).map_err(|e| Error::FileOpen(PathBuf::from(PROC_MOUNTS), e))?;
+
+        // Regex courtesy of Filippo.
+        // This will match on both v1 and v2 mount points.
+        // For v1, the "ver" capture group will be empty (len = 0)
+        let re = Regex::new(
+            r"^([a-z2]*)[[:space:]](?P<dir>.*)[[:space:]]cgroup(?P<ver>2?)[[:space:]](?P<options>.*)[[:space:]]0[[:space:]]0$",
+        ).map_err(Error::RegEx)?;
+
+        let mut found = false;
+        for l in BufReader::new(f).lines() {
+            let l = l.map_err(|e| Error::ReadLine(PathBuf::from(PROC_MOUNTS), e))?;
+            if let Some(capture) = re.captures(&l) {
+                if ver == CgroupVersion::V2 && capture["ver"].len() == 1 {
+                    // Found the cgroupv2 unified mountpoint; with cgroupsv2 there is only one
+                    // hierarchy so we insert it in the hashmap to use it later when creating
+                    // cgroups
+                    b.hierarchies
+                        .insert("unified".to_string(), PathBuf::from(&capture["dir"]));
+                    found = true;
+                    break;
+                } else if ver == CgroupVersion::V1 && capture["ver"].is_empty() {
+                    // Found a cgroupv1 mountpoint; with cgroupsv1 we can have multiple hiearachies.
+                    // Since we don't know which one will be used, we cache the mountpoints now,
+                    // and will create the hierachies on demand when a cgroup is built.
+                    b.mount_points.push(CgroupMountPoint {
+                        dir: String::from(&capture["dir"]),
+                        options: String::from(&capture["options"]),
+                    });
+                    found = true;
+                }
+            }
+        }
+        if found {
+            Ok(b)
+        } else {
+            Err(Error::CgroupHierachyMissing(
+                "No mount point found for this cgroup version.".to_string(),
+            ))
+        }
+    }
+
+    // Creates a new cggroup and returns it
+    pub fn new_cgroup(
+        &mut self,
+        file: String,
+        value: String,
+        id: &str,
+        exec_file_name: &OsStr,
+    ) -> Result<Box<dyn Cgroup>> {
```

Comment:
If we drop the Cgroup trait, I think the code becomes more simple to read, as I see there is not much shared between the 2 impls. We could have new_v1_cgroup and new_v2_cgroup.

Reply:
This was my initial thought also but while it would make code in cgroups.rs shorter, the code in env.rs would become more ugly for a few reasons:

- While impls for cgroup v1 and v2 differ, the interface for interacting with cgroups remains the same.

- Dropping the trait would require the Env struct to have 2 vectors to hold Cgroup{v1/v2} objects out of which only one would be used during the entire lifetime of the jailer.

- Because Env is the one that parses the cgroup-version argument it would have to have 2 branches that build objects that look the same for each version (same arguments to build). This would be true also for every usage of the objects (writing the value and attaching pids) which looks like a lot of boiler plate.

Reply:
I agree, makes sense.